File: CSE 2202/Lab_03/QuickHull_Algorithm.cpp
#include <bits/stdc++.h>
using namespace std;

#define point pair<int, int>
#define PI 3.141592654
set<point> Convex_Hull;

int getSide(point p1, point p2, point p)
{
    int x = (p.second - p1.second) * (p2.first - p1.first) - (p2.second - p1.second) * (p.first - p1.first);
    if (x < 0)
        return -1;
    if (x > 0)
        return 1;
    return 0;
}

int getDistance(point p1, point p2, point p)
{
    return abs((p2.first - p1.first) * (p.second - p1.second) - (p2.second - p1.second) * (p.first - p1.first));
}

void QuickHull(vector<point> p, point p1, point p2, int side)
{
    int max_dist = 0, index = -1;

    for (int i = 0; i < p.size(); i++)
    {
        int temp_dist = getDistance(p1, p2, p[i]);
        if (getSide(p1, p2, p[i]) == side and temp_dist > max_dist)
        {
            index = i;
            max_dist = temp_dist;
        }
    }
    if (index == -1)
    {
        Convex_Hull.insert(p1);
        Convex_Hull.insert(p2);
        return;
    }
    QuickHull(p, p[index], p1, -getSide(p1, p2, p[index]));
    QuickHull(p, p[index], p2, -getSide(p2, p1, p[index]));
}
void Calculate_Hull(vector<point> p)
{
    if (p.size() < 3)
    {
        cout << "Convex Hull can't be formed\n";
        return;
    }

    int min_x = 0, max_x = 0;

    for (int i = 1; i < p.size(); i++)
    {
        if (p[i].first < p[min_x].first)
            min_x = i;
        if (p[i].first > p[max_x].first)
            max_x = i;
    }
    QuickHull(p, p[min_x], p[max_x], 1);
    QuickHull(p, p[min_x], p[max_x], -1);
}
int main()
{
    vector<point> p, ans;
    int n;
    cin >> n;
    for (int i = 0; i < n; i++)
    {
        int x = rand() % 100, y = rand() % 100;
        p.push_back({x, y});
    }
    cout << "x=[";
    for (int i = 0; i < n; i++)
    {
        (i != n - 1) ? cout << p[i].first << "," : cout << p[i].first;
    }
    cout << "]\ny=[";
    for (int i = 0; i < n; i++)
    {
        (i != n - 1) ? cout << p[i].second << "," : cout << p[i].second;
    }
    cout << "]\n";

    Calculate_Hull(p);

    for (auto z : Convex_Hull)
        ans.push_back(z);

    int ymin = ans[0].second, min = 0;

    for (int i = 1; i < n; i++)
    {
        int y = ans[i].second;

        if ((y < ymin) || (ymin == y && ans[i].first < ans[min].first))
        {
            ymin = ans[i].second;
            min = i;
        }
    }
    swap(ans[min], ans[0]);

    vector<pair<double, pair<int, int>>> points;
    for (int i = 1; i < ans.size(); i++)
    {
        int x = ans[i].first, y = ans[i].second;
        double angle = (atan2(y, x) * 180 / PI);
        angle+=(angle<0)*360 ;
        points.push_back({angle, {x, y}});
    }
    sort(points.begin(), points.end());

    cout << "Points of Convex Hull are:\n";
    cout << "x=[";
    for (int i = 0; i < points.size(); i++)
    {
        cout << points[i].second.first << ",";
    }
    cout << points[0].second.first << "]\ny=[";
    for (int i = 0; i < points.size(); i++)
    {
        cout << points[i].second.second << ",";
    }
    cout << points[0].second.second << "]\n";
}

File: CSE 2202/Lab_03/Quick_Hull_Algorithm.py
import random
import time
import matplotlib.pyplot as plt

def Graph(points,s):
    x = [point[0] for point in points]
    y = [point[1] for point in points]
    if s == "scatter": plt.scatter(x,y,c='r')
    if s == "plot": plt.plot(x,y,c='b')

def getSide(p1,p2,p3):
    
    val =  (p3[1] - p2[1]) * (p2[0] - p1[0]) - (p2[1] - p1[1]) * (p3[0] - p2[0])
    
    if val>0 : # Anticlockwise
        return 1
    else:
        return -1 if val<0 else 0 # -1 -> Clockwise, 0 -> Co-linear

def getDistance(p1,p2,p3):
    return abs((p3[1] - p1[1]) * (p2[0] - p1[0]) - (p2[1] - p1[1]) * (p3[0] - p1[0]))

def QuickHull(points,p1,p2,side,hull):
    max_dist = 0
    max_point = []

    for point in points :
        temp_dist = getDistance(p1,point,p2)
        if getSide(p1,point,p2) == side and max_dist < temp_dist :
            max_dist = temp_dist
            max_point = point 

    if max_point :
        hull.append(p1)
        hull.append(max_point)
        hull.append(p2)

        QuickHull(points,p1,max_point,side,hull)
        QuickHull(points,max_point,p2,side,hull)

def Calculate_Hull():
    

    n=int(input())
    if n<3 :
        print(f'No convex hull can be formed')
        return

    points = [[random.randint(-150,150),random.randint(-150,150)] for _ in range (n)]

    min_x,max_x=0,0

    for i in range(n):
        if points[i][0] < points[min_x][0] : min_x = i
        if points[i][0] > points[max_x][0] : max_x = i

    hull1 = []
    hull2 = []
    start= time.perf_counter()

    QuickHull(points,points[min_x],points[max_x],1,hull1)
    QuickHull(points,points[min_x],points[max_x],-1,hull2)

    end = time.perf_counter()
    
    hull1.sort(reverse=True)
    hull2.sort()

    Convex_Hull = hull1 + hull2
    

    print('Time elapsed:',end-start,'seconds')

    print(f'Points of the convex hull are:')
    for point in Convex_Hull:
        print('('+str(point[0])+','+str(point[1])+')',end = ',')
    
    p=[points[min_x],points[max_x]]
    Graph(points,"scatter")
    Graph(p,"scatter")
    Graph(Convex_Hull,"plot")
    plt.show()

Calculate_Hull()